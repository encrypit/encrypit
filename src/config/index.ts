export const API_URL = import.meta.env.VITE_API_URL;
export const DEV = import.meta.env.DEV;
