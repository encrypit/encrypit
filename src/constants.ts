export const FILE = 'FILE';

export const ONE_MEGABYTE_IN_BYTES = 1048576;

export enum MAX_SIZE {
  DEFAULT = ONE_MEGABYTE_IN_BYTES * 5,
}

export enum MAX_FILES {
  DEFAULT = 1,
}

/**
 * @see {@link https://en.wikipedia.org/wiki/List_of_HTTP_status_codes}
 */
export enum HTTP_STATUS_CODES {
  LENGTH_REQUIRED = 411,
  NOT_FOUND = 404,
}
