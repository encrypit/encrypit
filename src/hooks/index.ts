import { fileApi } from '../store/api';

export const {
  useDeleteFileMutation,
  useUploadFileMutation,
  useLazyDownloadFileQuery,
} = fileApi;

export * from './useDispatch';
export * from './useSelector';
export * from './useSetDocumentTitle';
