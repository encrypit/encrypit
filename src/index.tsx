import '@fontsource/roboto/300.css';
import '@fontsource/roboto/400.css';
import '@fontsource/roboto/500.css';
import '@fontsource/roboto/700.css';

import { createRoot } from 'react-dom/client';

import App from './components/App';

export const root = createRoot(document.getElementById('root')!);

root.render(<App />);
