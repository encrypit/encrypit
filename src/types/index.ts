export interface User {
  email: string;
  id: string;
}
