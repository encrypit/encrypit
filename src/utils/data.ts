enum Events {
  error = 'error',
  loadend = 'loadend',
}

/**
 * Converts Blob to Base64.
 *
 * @param blob - Blob.
 * @returns - Base64.
 */
export async function blobToBase64(blob: Blob): Promise<string> {
  return new Promise((resolve, reject) => {
    const reader = new FileReader();

    function loadend() {
      resolve(reader.result as string);
      reader.removeEventListener(Events.loadend, loadend);
    }
    reader.addEventListener(Events.loadend, loadend);

    /* istanbul ignore next */
    function error() {
      reject(reader.error);
      reader.removeEventListener(Events.error, error);
    }
    reader.addEventListener(Events.error, error);

    reader.readAsDataURL(blob);
  });
}
