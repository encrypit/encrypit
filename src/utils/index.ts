export * from './filename';
export * from './form';
export * from './hash';
export * from './zip';
