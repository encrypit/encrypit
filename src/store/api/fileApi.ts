import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';
import { blobToBase64 } from 'file64';
import { HEADERS } from 'shared/constants';
import { API_URL } from 'src/config';
import type { DownloadFileResponse } from 'src/types';

export const fileApi = createApi({
  reducerPath: 'fileApi',

  baseQuery: fetchBaseQuery({
    baseUrl: `${API_URL}/api/files`,
  }),

  endpoints: (build) => ({
    /**
     * DELETE /api/files/[key]
     */
    deleteFile: build.mutation<void, string>({
      query: (fileKey: string) => ({
        url: `/${fileKey}`,
        method: 'DELETE',
      }),
    }),

    /**
     * GET /api/files/[key]
     */
    downloadFile: build.query<
      DownloadFileResponse,
      { key: string; passwordSHA512: string }
    >({
      query: ({ key, passwordSHA512 }) => ({
        url: `/${key}`,
        headers: {
          [HEADERS.PASSWORD_SHA512]: passwordSHA512,
        },
        responseHandler: async (response) =>
          blobToBase64(await response.blob()),
      }),

      transformResponse: /* istanbul ignore next */ (base64: string, meta) => ({
        file: base64,
        customMetadata: JSON.parse(
          meta!.response!.headers.get(HEADERS.CUSTOM_METADATA)!
        ) as DownloadFileResponse['customMetadata'],
      }),

      transformErrorResponse: ({ status }) => ({ status }),
    }),

    /**
     * POST /api/files
     */
    uploadFile: build.mutation<string, FormData>({
      query: (formData) => ({
        url: '',
        method: 'POST',
        body: formData,
        responseHandler: 'content-type',
      }),
    }),
  }),
});
