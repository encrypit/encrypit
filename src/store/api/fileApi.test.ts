import { act, renderHook } from '@testing-library/react';
import { createFormData } from 'src/utils';
import { fetchMock, mockFiles, wrapper } from 'test/helpers';

import { fileApi } from './fileApi';

const key = 'fileKey';
const passwordSHA512 = 'passwordSHA512';

beforeEach(() => {
  jest.clearAllMocks();
});

describe('deleteFile', () => {
  it('deletes file given key', async () => {
    const { result } = renderHook(() => fileApi.useDeleteFileMutation(), {
      wrapper,
    });

    await act(() => {
      const [deleteFile] = result.current;
      deleteFile(key);
    });

    expect(fetchMock).toBeCalledTimes(1);
    expect(fetchMock.mock.calls[0][0]).toMatchInlineSnapshot(`
      Request {
        "agent": undefined,
        "compress": true,
        "counter": 0,
        "follow": 20,
        "size": 0,
        "timeout": 0,
        Symbol(Body internals): {
          "body": null,
          "disturbed": false,
          "error": null,
        },
        Symbol(Request internals): {
          "headers": Headers {
            Symbol(map): {},
          },
          "method": "DELETE",
          "parsedURL": Url {
            "auth": null,
            "hash": null,
            "host": "localhost",
            "hostname": "localhost",
            "href": "http://localhost/api/files/fileKey",
            "path": "/api/files/fileKey",
            "pathname": "/api/files/fileKey",
            "port": null,
            "protocol": "http:",
            "query": null,
            "search": null,
            "slashes": true,
          },
          "redirect": "follow",
          "signal": AbortSignal {},
        },
      }
    `);
  });
});

describe('downloadFile', () => {
  describe('success', () => {
    beforeEach(() => {
      const blob = new Blob(['blob'], { type: 'text/plain' });
      fetchMock.mockResponseOnce(blob as unknown as string);
    });

    it('downloads file given key', async () => {
      await act(() => {
        renderHook(
          () => fileApi.useDownloadFileQuery({ key, passwordSHA512 }),
          { wrapper },
        );
      });
      expect(fetchMock).toBeCalledTimes(1);
      expect(fetchMock.mock.calls[0][0]).toMatchInlineSnapshot(`
        Request {
          "agent": undefined,
          "compress": true,
          "counter": 0,
          "follow": 20,
          "size": 0,
          "timeout": 0,
          Symbol(Body internals): {
            "body": null,
            "disturbed": false,
            "error": null,
          },
          Symbol(Request internals): {
            "headers": Headers {
              Symbol(map): {
                "X-Password-Sha512": [
                  "passwordSHA512",
                ],
              },
            },
            "method": "GET",
            "parsedURL": Url {
              "auth": null,
              "hash": null,
              "host": "localhost",
              "hostname": "localhost",
              "href": "http://localhost/api/files/fileKey",
              "path": "/api/files/fileKey",
              "pathname": "/api/files/fileKey",
              "port": null,
              "protocol": "http:",
              "query": null,
              "search": null,
              "slashes": true,
            },
            "redirect": "follow",
            "signal": AbortSignal {},
          },
        }
      `);
    });
  });

  describe('error', () => {
    beforeEach(() => {
      fetchMock.mockRejectOnce(new Error());
    });

    it('responds with null', async () => {
      await act(() => {
        const { result } = renderHook(
          () => fileApi.useDownloadFileQuery({ key, passwordSHA512 }),
          { wrapper },
        );
        expect(result.current).toBe(null);
      });
      expect(fetchMock).toBeCalledTimes(1);
    });
  });
});

describe('uploadFile', () => {
  it('uploads files', async () => {
    const { result } = renderHook(() => fileApi.useUploadFileMutation(), {
      wrapper,
    });

    const formData = await createFormData({
      file: mockFiles(1)[0],
      password: 'password',
      version: '1.0.0',
    });

    await act(async () => {
      const [uploadFile] = result.current;
      uploadFile(formData);
    });

    expect(fetchMock).toBeCalledTimes(1);
    expect(fetchMock.mock.calls[0][0]).toMatchInlineSnapshot(`
      Request {
        "agent": undefined,
        "compress": true,
        "counter": 0,
        "follow": 20,
        "size": 0,
        "timeout": 0,
        Symbol(Body internals): {
          "body": {
            "data": [
              91,
              111,
              98,
              106,
              101,
              99,
              116,
              32,
              70,
              111,
              114,
              109,
              68,
              97,
              116,
              97,
              93,
            ],
            "type": "Buffer",
          },
          "disturbed": false,
          "error": null,
        },
        Symbol(Request internals): {
          "headers": Headers {
            Symbol(map): {
              "Content-Type": [
                "text/plain;charset=UTF-8",
              ],
            },
          },
          "method": "POST",
          "parsedURL": Url {
            "auth": null,
            "hash": null,
            "host": "localhost",
            "hostname": "localhost",
            "href": "http://localhost/api/files",
            "path": "/api/files",
            "pathname": "/api/files",
            "port": null,
            "protocol": "http:",
            "query": null,
            "search": null,
            "slashes": true,
          },
          "redirect": "follow",
          "signal": AbortSignal {},
        },
      }
    `);
  });
});
