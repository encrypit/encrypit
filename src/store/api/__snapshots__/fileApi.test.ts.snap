// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`deleteFile deletes file given key and password hash 1`] = `
Request {
  "agent": undefined,
  "compress": true,
  "counter": 0,
  "follow": 20,
  "size": 0,
  "timeout": 0,
  Symbol(Body internals): {
    "body": null,
    "disturbed": false,
    "error": null,
  },
  Symbol(Request internals): {
    "headers": Headers {
      Symbol(map): {
        "X-Password-Sha512": [
          "cc7222b9eb5b430fc446942a1a7b85f9593328f9c42f3a343a9c698e5574fdc3d9b435f94dbc667f82786b1747bbb27abe76bfddc473fbc2145b8d2e3265114e",
        ],
        "accept": [
          "application/json",
        ],
      },
    },
    "method": "DELETE",
    "parsedURL": Url {
      "auth": null,
      "hash": null,
      "host": "localhost",
      "hostname": "localhost",
      "href": "http://localhost/api/files/fileKey",
      "path": "/api/files/fileKey",
      "pathname": "/api/files/fileKey",
      "port": null,
      "protocol": "http:",
      "query": null,
      "search": null,
      "slashes": true,
    },
    "redirect": "follow",
    "signal": AbortSignal {},
  },
}
`;

exports[`downloadFile success downloads file given key 1`] = `
Request {
  "agent": undefined,
  "compress": true,
  "counter": 0,
  "follow": 20,
  "size": 0,
  "timeout": 0,
  Symbol(Body internals): {
    "body": null,
    "disturbed": false,
    "error": null,
  },
  Symbol(Request internals): {
    "headers": Headers {
      Symbol(map): {
        "X-Password-Sha512": [
          "passwordSHA512",
        ],
      },
    },
    "method": "GET",
    "parsedURL": Url {
      "auth": null,
      "hash": null,
      "host": "localhost",
      "hostname": "localhost",
      "href": "http://localhost/api/files/fileKey",
      "path": "/api/files/fileKey",
      "pathname": "/api/files/fileKey",
      "port": null,
      "protocol": "http:",
      "query": null,
      "search": null,
      "slashes": true,
    },
    "redirect": "follow",
    "signal": AbortSignal {},
  },
}
`;

exports[`uploadFile uploads files 1`] = `
Request {
  "agent": undefined,
  "compress": true,
  "counter": 0,
  "follow": 20,
  "size": 0,
  "timeout": 0,
  Symbol(Body internals): {
    "body": {
      "data": [
        91,
        111,
        98,
        106,
        101,
        99,
        116,
        32,
        70,
        111,
        114,
        109,
        68,
        97,
        116,
        97,
        93,
      ],
      "type": "Buffer",
    },
    "disturbed": false,
    "error": null,
  },
  Symbol(Request internals): {
    "headers": Headers {
      Symbol(map): {
        "Content-Type": [
          "text/plain;charset=UTF-8",
        ],
      },
    },
    "method": "POST",
    "parsedURL": Url {
      "auth": null,
      "hash": null,
      "host": "localhost",
      "hostname": "localhost",
      "href": "http://localhost/api/files",
      "path": "/api/files",
      "pathname": "/api/files",
      "port": null,
      "protocol": "http:",
      "query": null,
      "search": null,
      "slashes": true,
    },
    "redirect": "follow",
    "signal": AbortSignal {},
  },
}
`;
